### ISSUE TITLE
External Payload Support for CoseSign1

### ISSUE THREAD

**Issue by alex-richards** [veraison/go-cose#195]
### What is the areas you would like to add the new feature to?

Go-COSE Library

### Is your feature request related to a problem?

It's currently not possible to verify the signature of an external payload.

### What solution do you propose?

Add a `Verify` that takes the payload as a parameter.

### What alternatives have you considered?

Setting `Payload` on the `CoseSign1` object . Not great, would refer not to modify the original object.
Constructing `SigStructure` validating with the key directly. `SigStructure` isn't exported.

**Comment by shizhMSFT** [comment]
> It's currently not possible to verify the signature of an external payload.

It is possible as you have considered the alternative

> Setting `Payload` on the `CoseSign1` object . Not great, would refer not to modify the original object.

Well, it is not great to modify the original object. However, you can modify the cloned object. And.. This is by design.

**Comment by shizhMSFT** [comment]
Consider the following example

```go
func VerifyP256(publicKey crypto.PublicKey, external, sig []byte) error {
    // create a verifier from a trusted private key
    verifier, err := cose.NewVerifier(cose.AlgorithmES256, publicKey)
    if err != nil {
        return err
    }

    // create a sign message from a raw COSE_Sign1 payload
    var msg cose.Sign1Message
    if err = msg.UnmarshalCBOR(sig); err != nil {
        return err
    }
    msg.Payload = external
    return msg.Verify(nil, verifier)
}
```

If the `msg` is not meant to be modified, you may consider
```go
func VerifyP256(publicKey crypto.PublicKey, msg *cose.Sign1Message, external []byte) error {
    // create a verifier from a trusted private key
    verifier, err := cose.NewVerifier(cose.AlgorithmES256, publicKey)
    if err != nil {
        return err
    }

    // clone the sign message to re-attach the content.
    msgToBeVerified := *msg
    msgToBeVerified.Payload = external
    return msgToBeVerified.Verify(nil, verifier)
}
```

**Comment by shizhMSFT** [comment]
Maybe all we need is to improve the doc like adding more examples?

**Comment by alex-richards** [comment]
Is "attach" documented, it doesn't appear in 1851 at all?

**Comment by shizhMSFT** [comment]
> Is "attach" documented, it doesn't appear in 1851 at all?

No, it is not documented in neither README.md nor the godoc. Do you want to write documentation? or I can send out a PR for them.

**Comment by shizhMSFT** [comment]
Nonetheless, I think we still need your PR #197 for the implementation of the following utilities
```go
func Sign1Detached(rand io.Reader, signer Signer, headers Headers, payload []byte, ...) ([]byte, error)
func Sign1UntaggedDetached(rand io.Reader, signer Signer, headers Headers, payload []byte, ...) ([]byte, error)
```

**Comment by shizhMSFT** [comment]
> > Is "attach" documented, it doesn't appear in 1851 at all?
> 
> No, it is not documented in neither README.md nor the godoc. Do you want to write documentation? or I can send out a PR for them.

I've created a PR #205 to add examples for detached payload in godoc.

**Comment by SteveLasker** [comment]
@alex-richards, can you weigh in on #205 to see if that unblocks #197?

**Comment by alex-richards** [comment]
@shizhMSFT Am I correct in thinking that you want to see all the Detached functions removed?

**Comment by shizhMSFT** [comment]
> @shizhMSFT Am I correct in thinking that you want to see all the Detached functions removed?

@alex-richards  Since I've written the missing docs in #205, I'd suggest removing all the detached functions except the below 2 utility functions in #197.

```go
func Sign1Detached(rand io.Reader, signer Signer, headers Headers, detached, external []byte) ([]byte, error)
func Sign1UntaggedDetached(rand io.Reader, signer Signer, headers Headers, detached, external []byte) ([]byte, error)
```

**Comment by SteveLasker** [comment]
@alex-richards, @shizhMSFT, gentle ping on Issue #195 and PR #197

**Comment by alex-richards** [comment]
@SteveLasker @shizhMSFT Not a decision I can make.

I need a concrete direction. 

Cheers, Alex

**Comment by SteveLasker** [comment]
Thanks, @alex-richards,
Do you feel comfortable with [Shiwei's suggestion](https://github.com/veraison/go-cose/issues/195#issuecomment-2350795836)?

**Comment by SteveLasker** [comment]
@alex-richards, if you have any concerns, please do let us know. This is a bidirectional conversation. Our goal is to complete your PR for #197 and then support #198

**Comment by alex-richards** [comment]
Hi @SteveLasker, my main concern is that this totally changes the intent of the PR. My goal was to add support for verifying detached payloads in a way that didn't introduce a load of fiddling with the internal members of the COSESign[1] structs.

@shizhMSFT's suggestion is to remove all of that, and replace it with a pair of utilities for creating SOSESign[1]s with detached payloads.

The question now is are the detached methods valuable or not? I've only seen one opinion on this point.

**Comment by OR13** [comment]
I do not think internal interfaces matter, except in maintainability cost.

What does matter is changing the external interfaces, and we do want to minimize those, so that the library is easy to use.

As I understand https://github.com/veraison/go-cose/issues/195#issuecomment-2350795836

The suggestion is to export 2 new detached sign convenience functions, and use the attached sign existing functions to satisfy them, reducing overall code.

I am in favor of this proposal.

**Comment by thomas-fossati** [comment]
I am ambivalent, with a slight preference for @alex-richards's approach, which I reckon has pretty good ergonomics.  I understand @shizhMSFT argument about code minimisation, but at the same time, I wonder if that's a false economy in that it does little to increase the usability of the API in "detached" scenarios.  My two cents, of course.

**Comment by shizhMSFT** [comment]
> What does matter is changing the external interfaces, and we do want to minimize those, so that the library is easy to use.

From security point of view, minimizing external interfaces minimizes the [attack surface](https://en.wikipedia.org/wiki/Attack_surface) so that the library has less change to have vulnerabilities.

**Comment by SteveLasker** [comment]
Hi folks,
I'd really like to unblock @alex-richards here. Can we please focus on completing these prs:
- #197 
- #198
